### ISSUE TITLE
Configuration beforeResolve does not allow further modification of the configuration

### ISSUE THREAD

**Issue by lacasseio** [gradle/gradle#15729]
### Expected Behavior

The configuration should only be locked after the `beforeResolve` actions are executed. When deferring the configuration as much as possible (aka using Provider API), it can be useful to have a last chance to modify before resolution.

### Current Behavior

When trying to change attributes of the configuration inside the `beforeResolve` actions, the error is confusing and wrong: `Cannot change attributes of dependency configuration ':test' after it has been resolved`. The configuration haven't been resolved yet as we are before the resolution. It seems the locking is happening too early leaving no room to defer configuration.

### Steps to Reproduce 

The following code is vastly simplified from what I have but the main ideas is represented:
```
def c = configurations.create('test')
c.incoming.beforeResolve { 
    it.attributes.attribute(Usage.USAGE_ATTRIBUTE, objects.named(Usage, 'some-usage')
}
c.resolve() // forcing, but any resolve will result in the same bug
```

### Your Environment

Build scan URL:

**Comment by lacasseio** [comment]
I see `ConfigurationInternal#beforeLocking` which seems to be more how I would expect `beforeResolve` to behave. I will use this as a workaround but there seems to be an public API gap here.

**Comment by ljacomet** [comment]
Would a lazy API on the attribute container help for your use case?

We are unlikely to open up more `beforeResolve` to mutations and instead close it up completely.

**Comment by lacasseio** [comment]
Not really, the issue is with the nesting of domain objects. Suppose `Configuration` child of `Component` and/or `Configuration` child of `Variant` which is a child of `Component`. In theory, we should realize the parent domain objects, i.e. the component and/or variant when the `Configuration` is realized. Unfortunately, the container doesn't play nice with such approaches as `Configuration` are often if not always realized which forces the realization of pretty much all domain objects as most of them to have some sort of consumable or resolvable `Configuration`. Ideally, Gradle would only realize each `Configuration` when they are searching for the "matching" `Configuration` and stop realizing further when the correct `Configuration` is found. It would allow users to properly configure the requested `Configuration` when configured actions are delayed through other means.

The `beforeLocking` sort-of fix helps with this use case by allowing the delayed actions to execute and provide the correct attributes, dependencies, etc. _before_ Gradle dependency engine kicks in and performs the resolution. It's not perfect as it will lock every configuration upon any dependency resolution but we can still prevent some configuration when a project doesn't participate in dependency resolution or when other tasks execute that don't resolve any dependencies. Note that widespread locking only happens when consuming a dependency. When resolving we are laser-focused on which `Configuration` to resolve.
